psp/src/vram_alloc.rs

```diff
@@ -0,0 +1,91 @@
+use crate::sys::TexturePixelFormat;
+use crate::sys::sceGeEdramGetSize;
+use core::mem::size_of;
+
+pub struct VramMemChunk {
+    start: u32,
+    len: u32,
+}
+
+impl VramMemChunk {
+    fn new(start: u32, len: u32) -> Self {
+        Self { start, len }
+    }
+
+    pub fn start(&self) -> u32 {
+        self.start
+    }
+
+    pub fn len(&self) -> u32 {
+        self.len
+    }
+}
+
+// A dead-simple VRAM bump allocator.
+pub struct SimpleVramAllocator {
+   offset: u32,
+}
+
+impl SimpleVramAllocator {
+    pub fn new() -> Self {
+        Self { offset: 0 }
+    }
+
+    pub fn alloc(&mut self, size: u32) -> VramMemChunk {
+        let old_offset = self.offset;
+        self.offset += size;
+
+        if self.offset > self.total_mem() {
+            panic!("Total VRAM size exceeded!");
+        }
+
+        VramMemChunk::new(old_offset, size)
+    }
+
+    pub fn dealloc(&mut self, _chunk: VramMemChunk) {
+        unimplemented!("Deallocation is not supported for the simple allocator.");
+    }
+
+    pub fn realloc(&mut self, _chunk: VramMemChunk) -> VramMemChunk {
+        unimplemented!("Reallocation is not supported for the simple allocator.");
+    }
+
+    pub fn alloc_sized<T: Sized>(&mut self, count: u32) -> VramMemChunk {
+        let size = size_of::<T>() as u32;
+        self.alloc(count * size)
+    }
+
+    pub fn alloc_texture_pixels(
+        &mut self,
+        width: u32,
+        height: u32,
+        psm: TexturePixelFormat,
+    ) -> VramMemChunk {
+        let size = get_memory_size(width, height, psm);
+        self.alloc(size)
+    }
+
+    fn total_mem(&self) -> u32 {
+        unsafe {
+            sceGeEdramGetSize()
+        }
+    }
+}
+
+// TODO: can we hardcode SCREEN_WIDTH and SCREEN_HEIGHT here, or is there a use-case
```

Comment:
Do you think we should leave this as width/height, or use SCREEN_WIDTH/SCREEN_HEIGHT? I'm not clear enough on the usecase for this one to know if it's likely to be used for smaller portions of the screen or not.

Reply:
Let's leave it as width and height, this allows support for loading textures into VRAM with arbitrary sizes